video/decode/dec_video.c

```diff
@@ -405,6 +405,17 @@ void video_work(struct dec_video *d_video)
     if (d_video->current_mpi || !d_video->vd_driver)
         return;
 
+    if (!d_video->packet && !d_video->new_segment) {
+        bool progress = receive_frame(d_video, &d_video->current_mpi);
+        if (!progress) {
+            d_video->current_state = DATA_EOF;
```

Comment:
I think I'd be more comfortable if this fell through to the old code below. Would also reduce the code size of this change. (Not fond of having two receive calls.)

Also, did you investigate whether it'd be possible to move all the receive_frame code simply before the send_frame one to avoid the duplication?

Once we have this in the final form, the same change should be made to dec_audio.c. (It's basically duplicated. I want to unify them one day, but for now it's not really possible yet.)

Reply:
The problem is the additional receive_frame() needs to happen before demux_read_packet_async(), which can block on I/O.

The usual flow is:

- demux_read_packet_async()
- send_packet()
- recieve_frame()

Swapping send/receive doesn't help because read is what blocks.

Reply:
I meant doing it like this:
- receive_frame()
- demux_read_packet_async()
- send_packet()

Not sure if the code structure easily allows this, but this sounds like it'd be optimal?

Reply:
Hmmm, that could work.

In the case where send_packet is called, no image would be returned until the next playloop iteration. But maybe that doesn't matter.

Reply:
Possibly you could change it to call the work function twice if there was no progress, or so. (Just rename the work function, and create a new video_work function that does that.)

The original intention might have been to require only 1 video_work call to feed a new packet and read a new frame to avoid redundant playloop iterations. (Because the playloop is getting too heavy.)

Reply:
video_work() is pretty big and I don't fully understand the framedrop or segment code, so I'm hesitant to make large changes. But I'll give it a try and see what happens.

Calling recieve_frame first makes sense in theory, and looping twice on DATA_AGAIN also makes sense. I'm still unsure of where the framedrop vdctrl would go in the new sequence, since it appears to use the next packet in the current implementation.

Reply:
The framedrop stuff has 2 purposes:
1. Specifically drop certain packets if precise seeking is used, and the packet has a PTS before the seek target (of course works only for frames that are not references, so vd_lavc.c uses AVDISCARD_NONREF for them). This is the reason why framedrop somehow depends on the input packet.
2. The old decoder framedrop mode, where the playloop tries to drop random frames in the decoder.

Since the send/receive API was introduced in libavcodec, it's not actually clear where the framedrop flag is read. With the old decode API, it always affected packet that was passed to the decode call. But with the new API it's not really clear which call actually decodes the packet and reads the framedrop. That's why the code tries to make sure to update the framedrop flags before the send and receive calls.

In theory, this should be fine:
1. "drain" the decoder of immediately buffered packets/frames by calling receive_frame until no more frames are returned
2. update the framedrop flags
3. call send_packet

This should work since the framedrop flags "stick" (both as far as vd_lavc.c and libavcodec are concerned). There's also no danger of setting the framedrop flags for the "wrong" packet, since at 2. all immediately buffered packets have been decoded and finished off.

I'm getting more and more convinced this is actually the right way. You want to reduce buffering as much as possible to reduce latency or stream stalls, and the old/current approach always reads a packet ahead for no good reason.

Reply:
> I'm getting more and more convinced this is actually the right way. You want to reduce buffering as much as possible to reduce latency or stream stalls, and the old/current approach always reads a packet ahead for no good reason.

I agree :+1:

The logic you suggested makes perfect sense, but I'm struggling to refactor the code to match. I hope you can take a look. Feel free to discard this PR and work on your own version if that's easier.